### PR TITLE
Auto backup

### PR DESCRIPTION
**Pull request by liuzhen**
1. add a dep of erlcron forked, removing the version dep (you can change it)
2. sent to test branch, for your review first.

### CODE REVIEW

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 1]
```diff
@@ -569,6 +605,79 @@ rolling_back_snapshot(next, State = #state{orig_state = NextState,
...
     {next_state, NextState, State#state{orig_state=undefined, args={}}}.
 
+doing_auto_backup_snapshot(next, State = #state{orig_state = NextState,
+                                             args={SnapID, JobName, Max}}) ->
+    lager:info("Auto-backup snapshot key: ~p Max: ~p", [JobName, Max]),
+    {ok, VMObj} = ls_vm:get(State#state.uuid),
+    ExistingSnapshots = ft_vm:snapshots(VMObj),
+    OutdatedAutoBackupSnapIDs = get_outdated_auto_backup_snapids(ExistingSnapshots, JobName, Max, []),
+    snapshot_action(State#state.uuid, SnapID, fun do_auto_backup_snapshot/4,
+                    fun finish_auto_backup_snapshot/4, [JobName, OutdatedAutoBackupSnapIDs]),
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: Line 615 is too long:                     fun finish_auto_backup_snapshot/4, [JobName, OutdatedAutoBackupSnapIDs]),.

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 2]
```diff
@@ -1637,3 +1819,39 @@ vm_event(UUID, Event) ->
...
+            OrderFun = fun({Timestamp1, _}, {Timestamp2, _}) ->
+                Timestamp1 =< Timestamp2
+            end,
+            OrderedAutoBackupSnapIDs = lists:sort(OrderFun, AutoBackupSnapIDs),
+            get_outdated_auto_backup_snapids(OrderedAutoBackupSnapIDs, OutdatedNum, []);
+        false ->
+            []
+    end.
+
+get_outdated_auto_backup_snapids([{_, SnapID}|T], OutdatedNum, OutdatedAutoBackupSnapIDs) when OutdatedNum > 0 ->
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: Line 1854 is too long: get_outdated_auto_backup_snapids([{_, SnapID}|T], OutdatedNum, OutdatedAutoBackupSnapIDs) when OutdatedNum > 0 ->.

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 3]
```diff
@@ -1637,3 +1819,39 @@ vm_event(UUID, Event) ->
...
+    end;
+get_outdated_auto_backup_snapids([], _, Max, AutoBackupSnapIDs) ->
+    OutdatedNum = erlang:length(AutoBackupSnapIDs) - Max + 1,
+    case OutdatedNum > 0 of
+        true ->
+            OrderFun = fun({Timestamp1, _}, {Timestamp2, _}) ->
+                Timestamp1 =< Timestamp2
+            end,
+            OrderedAutoBackupSnapIDs = lists:sort(OrderFun, AutoBackupSnapIDs),
+            get_outdated_auto_backup_snapids(OrderedAutoBackupSnapIDs, OutdatedNum, []);
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: Line 1849 is too long:             get_outdated_auto_backup_snapids(OrderedAutoBackupSnapIDs, OutdatedNum, []);.

File: apps/chunter/src/chunter_auto_backup.erl [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+            [{VM, Jobs}] ->
+                [{JobName, JobRef}|Jobs];
+            _ ->
+                [{JobName, JobRef}]
+        end,
+        ets:insert(?ETS_VM_JOBS, {VM, NewJobs}),
+        JobRef
+    catch
+        T:E ->
+            lager:error("[auto-backup] failed adding cronjob, reason:~p~n", [E]),
```
Comment (elvisci) [inline apps/chunter/src/chunter_auto_backup.erl]: Line 195 is too long:             lager:error("[auto-backup] failed adding cronjob, reason:~p~n", [E]),.

File: apps/chunter/src/chunter_auto_backup.erl [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+%%--------------------------------------------------------------------
+%% @doc
+%% Backup VM
+%%
+%% @spec backup(VM, JobName) -> ok
+%% @end
+%%--------------------------------------------------------------------
+backup(VM, <<"auto_backup_snap", _/binary>> = JobName) ->
+    Max = get_backup_max_num(VM, <<"auto_backup_max_snap">>),
+    lager:info("[auto-backup] Do backup: VM: ~p, JobName:~p Max:~p~n", [VM, JobName, Max]),
```
Comment (elvisci) [inline apps/chunter/src/chunter_auto_backup.erl]: Line 119 is too long:     lager:info("[auto-backup] Do backup: VM: ~p, JobName:~p Max:~p~n", [VM, JobName, Max]),.

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 6]
```diff
@@ -1142,6 +1296,34 @@ finish_rollback_snapshot(_VM, _SnapID, _, error) ->
     lager:error("Snapshot rollback failed!"),
     error.
 
+do_auto_backup_snapshot(<<_:1/binary, P/binary>>, VM, SnapID, [JobName, OutdatedAutoBackupSnapIDs]) ->
+    ls_vm:log(VM, <<"Creating auto-backup snapshot ", SnapID/binary, ": ", JobName/binary>>),
+    Result = chunter_zfs:snapshot(P, SnapID),
+
+    % delete outdated auto-backup snapshot
+    lists:foreach(fun(OutdatedAutoBackSnapID) ->
+        ls_vm:log(VM, <<"Deleting outdated auto-backup snapshot ", OutdatedAutoBackSnapID/binary, ": ", JobName/binary>>),
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: Line 1305 is too long:         ls_vm:log(VM, <<"Deleting outdated auto-backup snapshot ", OutdatedAutoBackSnapID/binary, ": ", JobName/binary>>),.

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 7]
```diff
@@ -1142,6 +1296,34 @@ finish_rollback_snapshot(_VM, _SnapID, _, error) ->
     lager:error("Snapshot rollback failed!"),
     error.
 
+do_auto_backup_snapshot(<<_:1/binary, P/binary>>, VM, SnapID, [JobName, OutdatedAutoBackupSnapIDs]) ->
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: Line 1299 is too long: do_auto_backup_snapshot(<<_:1/binary, P/binary>>, VM, SnapID, [JobName, OutdatedAutoBackupSnapIDs]) ->.

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 8]
```diff
@@ -1142,6 +1296,34 @@ finish_rollback_snapshot(_VM, _SnapID, _, error) ->
...
+    ls_vm:log(VM, <<"Creating auto-backup snapshot ", SnapID/binary, ": ", JobName/binary>>),
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: Line 1300 is too long:     ls_vm:log(VM, <<"Creating auto-backup snapshot ", SnapID/binary, ": ", JobName/binary>>),.

File: apps/chunter/src/chunter_auto_backup.erl [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+
+-module(chunter_auto_backup).
+
+-behaviour(gen_server).
+
+% gen_server APIs
+-export([start_link/0]).
+
+% gen_server callbacks
+-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2, code_change/3]).
```
Comment (elvisci) [inline apps/chunter/src/chunter_auto_backup.erl]: Line 15 is too long: -export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2, code_change/3])..

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 10]
```diff
@@ -927,6 +1073,14 @@ handle_sync_event(delete, _From, StateName, State) ->
             {reply, ok, StateName, State}
     end;
 
+handle_sync_event({auto_backup, snapshot, JobName, Max}, _From, StateName, State) ->
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: Line 1076 is too long: handle_sync_event({auto_backup, snapshot, JobName, Max}, _From, StateName, State) ->.

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 11]
```diff
@@ -1142,6 +1296,34 @@ finish_rollback_snapshot(_VM, _SnapID, _, error) ->
...
+
+    % delete outdated auto-backup snapshot
+    lists:foreach(fun(OutdatedAutoBackSnapID) ->
+        ls_vm:log(VM, <<"Deleting outdated auto-backup snapshot ", OutdatedAutoBackSnapID/binary, ": ", JobName/binary>>),
+        chunter_zfs:destroy_snapshot(P, OutdatedAutoBackSnapID, [f, r])
+    end, OutdatedAutoBackupSnapIDs),
+
+    Result.
+
+finish_auto_backup_snapshot(VM, SnapID, [JobName, OutdatedAutoBackupSnapIDs], ok) ->
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: Line 1311 is too long: finish_auto_backup_snapshot(VM, SnapID, [JobName, OutdatedAutoBackupSnapIDs], ok) ->.

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 12]
```diff
@@ -776,6 +885,31 @@ handle_event(update_fw, StateName,
...
+    AutoBackupFun =
+        fun(KV) ->
+                {Key, Value} = KV,
+                case Key of
+                    <<"auto_backup_", _/binary>> ->
+                        case chunter_auto_backup:add(UUID, Key, Value) of
+                            {error, _Reason} ->
+                                lager:error("Invalid auto backup rule: ~s ~s", [Key, Value]),
+                                ls_vm:set_metadata(UUID, [{Key, delete}]),
+                                ls_vm:set_metadata(UUID, [{<<"invalid_", Key/binary>>, Value}]);
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: Line 903 is too long:                                 ls_vm:set_metadata(UUID, [{<<"invalid_", Key/binary>>, Value}]);.

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 13]
```diff
@@ -1637,3 +1819,39 @@ vm_event(UUID, Event) ->
...
+                      {ok, T2} = jsxd:get(<<"timestamp">>, P2),
+                      T1 > T2
+              end,
+    lists:sort(SortFun, Backups).
+
+get_outdated_auto_backup_snapids([{SnapID, Attrs}|T], JobName, Max, AutoBackupSnapIDs) ->
+    case lists:keyfind(<<"comment">>, 1, Attrs) of
+        {<<"comment">>, JobName} ->
+            Timestamp = jsxd:get(<<"timestamp">>, 0, Attrs),
+            get_outdated_auto_backup_snapids(T, JobName, Max, [{Timestamp, SnapID}|AutoBackupSnapIDs]);
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: Line 1837 is too long:             get_outdated_auto_backup_snapids(T, JobName, Max, [{Timestamp, SnapID}|AutoBackupSnapIDs]);.

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 14]
```diff
@@ -776,6 +885,31 @@ handle_event(update_fw, StateName,
...
+    {ok, VmData} = ls_vm:get(UUID),
+    Meta = ft_vm:metadata(VmData),
+    AutoBackupFun =
+        fun(KV) ->
+                {Key, Value} = KV,
+                case Key of
+                    <<"auto_backup_", _/binary>> ->
+                        case chunter_auto_backup:add(UUID, Key, Value) of
+                            {error, _Reason} ->
+                                lager:error("Invalid auto backup rule: ~s ~s", [Key, Value]),
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: Line 901 is too long:                                 lager:error("Invalid auto backup rule: ~s ~s", [Key, Value]),.

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 15]
```diff
@@ -1142,6 +1296,34 @@ finish_rollback_snapshot(_VM, _SnapID, _, error) ->
...
+
+finish_auto_backup_snapshot(VM, SnapID, [JobName, OutdatedAutoBackupSnapIDs], ok) ->
+    lager:debug("Creating auto-backup snapshot:~p", [SnapID]),
+    ls_vm:set_snapshot(VM,
+                 [{[SnapID, <<"timestamp">>], timestamp()},
+                  {[SnapID, <<"comment">>], JobName},
+                  {[SnapID, <<"state">>], <<"completed">>}]),
+
+    lists:foreach(fun(OutdatedAutoBackSnapID) ->
+        lager:debug("Deleting outdated auto-backup snapshot:~p", [OutdatedAutoBackSnapID]),
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: Line 1319 is too long:         lager:debug("Deleting outdated auto-backup snapshot:~p", [OutdatedAutoBackSnapID]),.

File: apps/chunter/src/chunter_auto_backup.erl [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+%% @end
+%%--------------------------------------------------------------------
+backup(VM, <<"auto_backup_snap", _/binary>> = JobName) ->
+    Max = get_backup_max_num(VM, <<"auto_backup_max_snap">>),
+    lager:info("[auto-backup] Do backup: VM: ~p, JobName:~p Max:~p~n", [VM, JobName, Max]),
+    chunter_vm_fsm:auto_backup_snapshot(VM, JobName, Max),
+    ok;
+backup(VM, <<"auto_backup_full_", _/binary>> = JobName) ->
+    Max = get_backup_max_num(VM, <<"auto_backup_max_full">>),
+    lager:info("[auto-backup] Do backup: VM: ~p, JobName:~p Max:~p~n", [VM, JobName, Max]),
```
Comment (elvisci) [inline apps/chunter/src/chunter_auto_backup.erl]: Line 124 is too long:     lager:info("[auto-backup] Do backup: VM: ~p, JobName:~p Max:~p~n", [VM, JobName, Max]),.

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 17]
```diff
@@ -1637,3 +1819,39 @@ vm_event(UUID, Event) ->
...
+    SortFun = fun (B1, B2) ->
+                      {_, P1} = B1,
+                      {_, P2} = B2,
+                      {ok, T1} = jsxd:get(<<"timestamp">>, P1),
+                      {ok, T2} = jsxd:get(<<"timestamp">>, P2),
+                      T1 > T2
+              end,
+    lists:sort(SortFun, Backups).
+
+get_outdated_auto_backup_snapids([{SnapID, Attrs}|T], JobName, Max, AutoBackupSnapIDs) ->
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: Line 1833 is too long: get_outdated_auto_backup_snapids([{SnapID, Attrs}|T], JobName, Max, AutoBackupSnapIDs) ->.

File: apps/chunter/src/chunter_auto_backup.erl [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+%% @spec do_clear_all(List) -> ok
+%% @end
+%%--------------------------------------------------------------------
+do_clear_all(VM, [{JobName, JobRef}|T]) ->
+    try
+        erlcron:cancel(JobRef),
+        lager:info("[auto-backup] succeed to clear cronjob VM:~p JobName:~p~n", [VM, JobName])
+    catch
+        T:E ->
+            lager:error("[auto-backup] failed to clear cronjob VM:~p Jobname:~p JobRef: ~p, reason: ~p", [VM, JobName, JobRef, E]),
```
Comment (elvisci) [inline apps/chunter/src/chunter_auto_backup.erl]: Line 213 is too long:             lager:error("[auto-backup] failed to clear cronjob VM:~p Jobname:~p JobRef: ~p, reason: ~p", [VM, JobName, JobRef, E]),.

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 19]
```diff
@@ -569,6 +605,79 @@ rolling_back_snapshot(next, State = #state{orig_state = NextState,
...
+                case proplists:get_value(<<"comment">>, Props) of
+                    Key -> true;
+                    _ -> false
+                end
+        end,
+    case lists:filter(FilterIncFun, Backups) of
+        [] ->
+            case sort_backup(Backups) of
+                [] ->
+                    lager:warning("No usable parent backup, creating a full one.", []),
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: Line 662 is too long:                     lager:warning("No usable parent backup, creating a full one.", []),.

File: apps/chunter/src/chunter_auto_backup.erl [hunk 20]
```diff
@@ -0,0 +1,231 @@
...
+%% @doc
+%% Traversing all vm jobs and cancel them
+%%
+%% @spec do_clear_all(List) -> ok
+%% @end
+%%--------------------------------------------------------------------
+do_clear_all(VM, [{JobName, JobRef}|T]) ->
+    try
+        erlcron:cancel(JobRef),
+        lager:info("[auto-backup] succeed to clear cronjob VM:~p JobName:~p~n", [VM, JobName])
```
Comment (elvisci) [inline apps/chunter/src/chunter_auto_backup.erl]: Line 210 is too long:         lager:info("[auto-backup] succeed to clear cronjob VM:~p JobName:~p~n", [VM, JobName]).

File: apps/chunter/src/chunter_auto_backup.erl [hunk 21]
```diff
@@ -0,0 +1,231 @@
...
+    ets:new(?ETS_VM_JOBS, [named_table, set, public, {write_concurrency, true}]),
+
+    % delete all outdated cron jobs
+    OutdatedJobs = ecrn_reg:get_all(),
+    DeleteOutdatedJobs = fun({JobRef, _}) ->
+            erlcron:cancel(JobRef)
+    end,
+    lists:foreach(DeleteOutdatedJobs, OutdatedJobs),
+
+    % when this process is restarting, it calls every vm_fsm process to register cron job
```
Comment (elvisci) [inline apps/chunter/src/chunter_auto_backup.erl]: Line 48 is too long:     % when this process is restarting, it calls every vm_fsm process to register cron job.

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 22]
```diff
@@ -776,6 +885,31 @@ handle_event(update_fw, StateName,
...
+    lager:info("[vm:~s] set_metadata", [UUID]),
+    chunter_auto_backup:clear_all(UUID),
+    {ok, VmData} = ls_vm:get(UUID),
+    Meta = ft_vm:metadata(VmData),
+    AutoBackupFun =
+        fun(KV) ->
+                {Key, Value} = KV,
+                case Key of
+                    <<"auto_backup_", _/binary>> ->
+                        case chunter_auto_backup:add(UUID, Key, Value) of
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: The expression on line 899 and column 25 is nested beyond the maximum level of 3.

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 23]
```diff
@@ -569,6 +605,79 @@ rolling_back_snapshot(next, State = #state{orig_state = NextState,
                     fun finish_rollback_snapshot/4, []),
     {next_state, NextState, State#state{orig_state=undefined, args={}}}.
 
+doing_auto_backup_snapshot(next, State = #state{orig_state = NextState,
+                                             args={SnapID, JobName, Max}}) ->
+    lager:info("Auto-backup snapshot key: ~p Max: ~p", [JobName, Max]),
+    {ok, VMObj} = ls_vm:get(State#state.uuid),
+    ExistingSnapshots = ft_vm:snapshots(VMObj),
+    OutdatedAutoBackupSnapIDs = get_outdated_auto_backup_snapids(ExistingSnapshots, JobName, Max, []),
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: Line 613 is too long:     OutdatedAutoBackupSnapIDs = get_outdated_auto_backup_snapids(ExistingSnapshots, JobName, Max, []),.

File: apps/chunter/src/chunter_auto_backup.erl [hunk 24]
```diff
@@ -0,0 +1,231 @@
...
+%%
+%% @spec add(VM, JobName, When) -> {ok | error, Reason}
+%% @end
+%%--------------------------------------------------------------------
+
+do_cron_job(VM, JobName, When) ->
+    Job = {When, {?MODULE, backup, [VM, JobName]}},
+    try
+        JobRef = erlcron:cron(Job),
+        lager:info("[auto-backup] succeed adding cronjob, VM:~p JobName:~p, JobRef:~p", [VM, JobName, JobRef]),
```
Comment (elvisci) [inline apps/chunter/src/chunter_auto_backup.erl]: Line 183 is too long:         lager:info("[auto-backup] succeed adding cronjob, VM:~p JobName:~p, JobRef:~p", [VM, JobName, JobRef]),.

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 25]
```diff
@@ -1637,3 +1819,39 @@ vm_event(UUID, Event) ->
...
+                Timestamp1 =< Timestamp2
+            end,
+            OrderedAutoBackupSnapIDs = lists:sort(OrderFun, AutoBackupSnapIDs),
+            get_outdated_auto_backup_snapids(OrderedAutoBackupSnapIDs, OutdatedNum, []);
+        false ->
+            []
+    end.
+
+get_outdated_auto_backup_snapids([{_, SnapID}|T], OutdatedNum, OutdatedAutoBackupSnapIDs) when OutdatedNum > 0 ->
+    get_outdated_auto_backup_snapids(T, OutdatedNum - 1, [SnapID|OutdatedAutoBackupSnapIDs]);
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: Line 1855 is too long:     get_outdated_auto_backup_snapids(T, OutdatedNum - 1, [SnapID|OutdatedAutoBackupSnapIDs]);.

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 26]
```diff
@@ -569,6 +605,79 @@ rolling_back_snapshot(next, State = #state{orig_state = NextState,
...
+                    Key -> true;
+                    _ -> false
+                end
+        end,
+    case lists:filter(FilterIncFun, Backups) of
+        [] ->
+            case sort_backup(Backups) of
+                [] ->
+                    lager:warning("No usable parent backup, creating a full one.", []),
+                    preparing_auto_backup_full(next, State#state{args = {Key, 3}}),
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: Line 663 is too long:                     preparing_auto_backup_full(next, State#state{args = {Key, 3}}),.

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 27]
```diff
@@ -222,6 +246,17 @@ rollback_snapshot(UUID, SnapID) ->
     gen_fsm:sync_send_all_state_event({global, {vm, UUID}},
                                       {snapshot, rollback, SnapID}).
 
+auto_backup_snapshot(UUID, JobName, Max) ->
+    gen_fsm:sync_send_all_state_event({global, {vm, UUID}}, {auto_backup, snapshot, JobName, Max}).
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: Line 250 is too long:     gen_fsm:sync_send_all_state_event({global, {vm, UUID}}, {auto_backup, snapshot, JobName, Max})..

File: apps/chunter/src/chunter_auto_backup.erl [hunk 28]
```diff
@@ -0,0 +1,231 @@
...
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+
+
+%% ===================================================================
+%% gen_server callbacks
+%% ===================================================================
+
+init([]) ->
+    ets:new(?ETS_VM_JOBS, [named_table, set, public, {write_concurrency, true}]),
```
Comment (elvisci) [inline apps/chunter/src/chunter_auto_backup.erl]: Line 39 is too long:     ets:new(?ETS_VM_JOBS, [named_table, set, public, {write_concurrency, true}]),.

File: apps/chunter/src/chunter_vm_fsm.erl [hunk 29]
```diff
@@ -569,6 +605,79 @@ rolling_back_snapshot(next, State = #state{orig_state = NextState,
...
+                    _ -> false
+                end
+        end,
+    case lists:filter(FilterFun, ft_vm:backups(VmSpec)) of
+        TheBackups when length(TheBackups) > Max ->
+            Sorted = sort_backup(TheBackups),
+            lists:foreach(
+              fun(Old) ->
+                    {UUID, _} = Old,
+                    lager:info("Auto-backup deleting VM:~p key: ~p, BackupUUID:~p", [VM, Key, UUID]),
```
Comment (elvisci) [inline apps/chunter/src/chunter_vm_fsm.erl]: Line 637 is too long:                     lager:info("Auto-backup deleting VM:~p key: ~p, BackupUUID:~p", [VM, Key, UUID]),.

File: apps/chunter/src/chunter_auto_backup.erl [hunk 30]
```diff
@@ -0,0 +1,231 @@
...
+    catch
+        T:E ->
+            {T, E}
+    end;
+set_backup_max_num(VM, Type, MaxNum) when is_integer(MaxNum) ->
+    do_set_backup_max_num(VM, Type, MaxNum).
+
+do_set_backup_max_num(VM, Type, MaxNum) ->
+    ets:insert(?ETS_VM_JOBS, {{VM, Type}, MaxNum}),
+    lager:info("[auto-backup] set backup number:~p Type: ~p VM:~p~n", [MaxNum, Type, VM]),
```
Comment (elvisci) [inline apps/chunter/src/chunter_auto_backup.erl]: Line 167 is too long:     lager:info("[auto-backup] set backup number:~p Type: ~p VM:~p~n", [MaxNum, Type, VM]),.

**Comment by Licenser** [comment]
I am not a big fan of including a auto backup feature, it puts a lot of responsibility into FiFo's hands and I don't know if we want to shoulder this.

That aside as it is it's not mergable, it goes from dev -> test and with all the clutter it's near impossible to review.

All that said, putting the backup logic in chunter is the sanest approach I've seen so far and one that didn't even came to mind when I thought about it. If you clean the PR up a bit I'll look it over and we can see if it's safe enough to include.

**Pull request closed without merging**